### ISSUE TITLE
Remove boost dependency

### ISSUE THREAD

**Issue by youaresoomean** [IfcOpenShell/IfcOpenShell#125]
Hi,

apart from OpenCascade for IfcGeom boost is a huge dependency required for IfcOpenShell (400Mb). While OpenCascade is undoubtfully required and provides crucial functioanlity, boost does not, since very few boost libs are used (mainly programargs,regex and optional). All of those boost libs could easily be replaced with std::(regex) or some lightweight alternatives. 

Alternatively rely on header-only parts and pack them with IfcOpenShell.

**Comment by Stinkfist0** [comment]
Although I somewhat agree with you, this would be quite an undertaking. It would also restrict us to support only the most recent compilers and standards (e.g. boost/std::optional which is heavily used will be part of C++17). Personally I would love to jump to using the more recent C++ standards, but I can understand if C++03 conformance is wanted for the project for now.

**Comment by aothms** [comment]
In addition, can I ask what you would hope to accomplish by removing boost as a dependency? Reducing compile time? Reducing binary footprint by offloading more things to the stdlib?

Your listing is not entirely complete. We also rely heavily on `variant` and `dynamic_bitset`, some string manipulation for example and I can foresee `thread` being of some use in the near future. I must admit I don't follow the recent C++ developments in much detail so I am not fully aware which of these are now incorporated into the standard. As OCCT7 is requiring C++11, there is little reason not to start embracing it. We can do that starting from 0.6 and keep 0.5 around for older platforms.

But honestly I don't really see the benefit of removing boost. Isn't it something that most C++ programmers have on their system anyway? And besides, considering the dependency on OCCT, IfcOpenShell is never going to be a light-weight application. Curious to hear your perspectives.

**Comment by youaresoomean** [comment]
My issue with boost (or in general any additional dependency) is maintenance. When building for several platforms every update has it's cost (either compiler or lib update). Binaries have to get into source control/build servers and so on to ensure reproducible builds, DLLs have to be packed with the product etc.

Binary size and slow compile times are also an issue as already stated by @aothms above

**Comment by aothms** [comment]
Thanks for your reference on Boost BCP, I wasn't aware of this. Not so much in favor of using the optional/variant alternatives, but these are header only in Boost so can be easily isolated. This indeed seems to be an option combined with thread and regex from c++11. Since OCCT 7 is also C++11 I propose to switch to C++11 from the next release, 0.6.0.

boost::program_options is then still something that needs a replacement. But we are already running into some quirks pertaining to our usage of it in IfcConvert. So this might be worthwhile anyway.

So to summarize, from 0.6.0 onwards:
- C++11
- Use only header-only boost libs
- Isolate and ship them with Boost BCP

Does this sound like a plan? @youaresoomean We are still busy wrapping up 0.5.0, but naturally by that time you are more then welcome to pitch in.

**Comment by youaresoomean** [comment]
@aothms Perfect. I think it's a good and practicable solution.

**Comment by Moult** [comment]
Bump? Curious on what has happened since.

**Comment by aothms** [comment]
Well, since then, I think usage of boost has mostly been increasing, recently we started using Boost Preprocessor to iterate over the required schema version passed from cmake. We're also relying on a couple of string manipulation functions.

I think the it's still possible to separate out the header-only ones and include them in the repo and use C++11/14/17 features as substitutes for some of the boost libs that require compilation (`std::optional` is also finally a thing now, as well as `std::tuple`).

For me this is not a high priority issue though, we have bigger and more esoteric dependencies to worry about than boost. And compilation is pretty seamless due to all the build scripts.

I will rewrite the issue.

**Comment by BIMvoice** [comment]
Reading the thread, the last maintainer comment already settles the direction: boost usage has grown since this was filed (Boost Preprocessor now iterates schema versions passed from CMake, among others), so eliminating non-header-only boost is the opposite of where the codebase went. Recommend closing as wontfix to reflect that reality.
